# Exercise 4.5 Importing libraries

In [3]:
import pandas as pd
import numpy as np
import os

# Import Datasets

In [4]:
path = r'C:\Users\stefa\15-02-2022 Instacart Basket Analysis'


In [5]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [6]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [7]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week    order_time  days_since_prior_order  
count        3.421083e+06  3.421083e+06            3.214874e+06  
mean         2.776219e+00  1.345202e+01            1.111484e+01  
std          2.046829e+00  4.226088e+00            9.206737e+00  
min          0.000000e+00  0.000000e+00            0.000000e+00  
25%          1.000000e+00  1.000000e+01            4.000000e+00  
50%          3.000000e+00  1.300000e+01            7.000000e+00  
75%          5.000000e+00  1.600000e+01            1.500000e+01  
max          6.000000e+00  2.300000e+01            3.000000e+01

In [8]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [9]:
# Create a Dataframe
df_test = pd.DataFrame()

In [10]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [11]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [13]:
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [14]:
df_test['mix'] = df_test['mix'].astype('str')

In [16]:
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

In [17]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [18]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [19]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [30]:
df_nan.describe()

product_id    aisle_id  department_id     prices
count     16.000000   16.000000      16.000000  16.000000
mean    6684.000000   89.937500      10.937500  13.012500
std    12836.665242   33.731229       4.639953   3.881731
min       34.000000   26.000000       1.000000   1.200000
25%      459.250000   70.750000       7.750000  12.175000
50%     2413.000000   98.500000      11.500000  13.650000
75%     3872.750000  120.000000      14.500000  14.425000
max    40440.000000  126.000000      16.000000  20.900000

In [34]:
df_prods.shape

(49693, 5)

In [35]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [36]:
df_prods_clean.shape

(49677, 5)

In [37]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [38]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [41]:
df_prods_clean.shape

(49677, 5)

In [42]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()


In [44]:
df_prods_clean_no_dups.shape

(49672, 5)

In [45]:
# Exporting Dataframe as products_checked.csv
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# Directions 4.5

In [ ]:
# 1 Consistency checks are done during the Exercise

In [ ]:
# 2 Run the df.describe() function on your df_prods dataframe.

In [46]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

prices max 99999 looks off and should be investigated.

In [ ]:
# 3 Check for mixed-type data in your df_ords dataframe.

In [48]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

4.0 no mixed types have been found

In [ ]:
# 5 Run a check for missing values in your df_ords dataframe

In [49]:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_time                     0
days_since_prior_order    206209
dtype: int64

We have 206209 missing Values in the days_since_prior_order column. 

In [50]:
df_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [51]:
df_nan

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_time  days_since_prior_order  
0                 8                     NaN  
11               11                     NaN  
26               14                     NaN  
39               11                     NaN  
45               12                     NaN  
...             ...                     ...  
3420930          12                     NaN  
3420934          18                     NaN  
3421002          19                     NaN  
3421019          15                     NaN  
3421069          11                     NaN  

[206209 rows x 7 columns]

This means that the missing value 206209 for days_since_prior_order could be related to a customer who simply has purchased just one order. Therefore, no prior order is possible. This can be doublecheck and confirmed via a filter for each customer.

In [ ]:
# 6 Address the missing values using an appropriate method.

For this case, since there are customers who have no prior order, we cannot impute the values. We can however not simply remove/ filter the missing Data as it is also important data to know which customers have only one i.e the First Order. For this Case we should Create a new Variable 

In [52]:
df_ords_clean = df_ords

In [53]:
df_ords_clean['First_order'] = df_ords['days_since_prior_order'].isnull() == True

This should generate a neww column which indicates in each line a true or false if its a customer first oder or not.
We can test this with the describe function.

In [54]:
df_ords_clean.head(10)

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           1   2398795        1             2                   3   
2           2    473747        1             3                   3   
3           3   2254736        1             4                   4   
4           4    431534        1             5                   4   
5           5   3367565        1             6                   2   
6           6    550135        1             7                   1   
7           7   3108588        1             8                   1   
8           8   2295261        1             9                   1   
9           9   2550362        1            10                   4   

   order_time  days_since_prior_order  First_order  
0           8                     NaN         True  
1           7                    15.0        False  
2          12                    21.0        False  
3           7                    29.0        False  
4          15                    28.0        False  
5           7                    19.0        False  
6           9                    20.0        False  
7          14                    14.0        False  
8          16                     0.0        False  
9           8                    30.0        False

In [ ]:
# 7 Run a check for duplicate values in your df_ords data.

In [55]:
df_dups = df_ords_clean[df_ords_clean.duplicated()]

In [56]:
df_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_time, days_since_prior_order, First_order]
Index: []

After checking with the command above, we could see that no duplicates have been found.

In [ ]:
# Address the duplicates using an appropriate method.

In [57]:
df_dups.shape

(0, 8)

As identified in the previus Step, we have not found any duplicate Values and hence don't need to adjust anything.

In [ ]:
# Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.

In [59]:
df_ords_clean.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_cleaned.csv'))

In [60]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_cleaned.csv'))